tests/test_requests_compat.py

```diff
@@ -0,0 +1,2049 @@
+        assert request.url == expected
+
+    def test_params_original_order_is_preserved_by_default(self):
+        param_ordered_dict = collections.OrderedDict(
+            (("z", 1), ("a", 1), ("k", 1), ("d", 1))
+        )
+        request = requests.Request(
+            "GET", "http://example.com/", params=param_ordered_dict
+        )
+        assert request.url == "http://example.com/?z=1&a=1&k=1&d=1"
+
+    @pytest.mark.xfail(reason="HTTPX doesn't support bytes-typed 'params' yet")
+    def test_params_bytes_are_encoded(self):
+        request = requests.Request("GET", "http://example.com", params=b"test=foo")
+        assert request.url == "http://example.com/?test=foo"
+
+    def test_binary_put(self):
+        request = requests.Request(
+            "PUT", "http://example.com", data="ööö".encode("utf-8")
+        )
+        assert isinstance(request.content, bytes)
+
+    @pytest.mark.xfail(reason="Not supported by HTTPX yet", raises=InvalidURL)
+    def test_whitespaces_are_removed_from_url(self):
+        # Test for issue #3696
+        request = requests.Request("GET", " http://example.com")
+        assert request.url == "http://example.com/"
+
+    @pytest.mark.parametrize("scheme", ("http://", "HTTP://", "hTTp://", "HttP://"))
+    def test_mixed_case_scheme_acceptable(self, httpbin, scheme):
+        from urllib.parse import urlparse
+
+        s = requests.Session()
+        parts = urlparse(httpbin("get"))
+        url = scheme + parts.netloc + parts.path
+        r = s.request("GET", url)
+        assert r.status_code == 200, "failed for scheme {}".format(scheme)
+
+    def test_HTTP_200_OK_GET_ALTERNATIVE(self, httpbin):
+        s = requests.Session()
+        r = s.request("GET", httpbin("get"))
+        assert r.status_code == 200
+
+    def test_HTTP_302_ALLOW_REDIRECT_GET(self, httpbin):
+        r = requests.get(httpbin("redirect", "1"))
+        assert r.status_code == 200
+        assert r.history[0].status_code == 302
+        assert r.history[0].is_redirect
+
+    def test_HTTP_307_ALLOW_REDIRECT_POST(self, httpbin):
+        r = requests.post(
+            httpbin("redirect-to"),
+            data="test",
+            params={"url": "post", "status_code": 307},
+        )
+        assert r.status_code == 200
+        assert r.history[0].status_code == 307
+        assert r.history[0].is_redirect
+        assert r.json()["data"] == "test"
+
+    @pytest.mark.xfail(reason="Receives 501 instead of 200")
+    def test_HTTP_307_ALLOW_REDIRECT_POST_WITH_SEEKABLE(self, httpbin):
+        byte_str = b"test"
+        r = requests.post(
+            httpbin("redirect-to"),
+            data=io.BytesIO(byte_str),
+            params={"url": "post", "status_code": 307},
+        )
+        assert r.status_code == 200
+        assert r.history[0].status_code == 307
+        assert r.history[0].is_redirect
+        assert r.json()["data"] == byte_str.decode("utf-8")
+
+    @pytest.mark.xfail(
```

Comment:
Was there a reason this number was chosen over 20? Both are huge :) @tomchristie

Reply:
I'm not clear from the context what we're talking about here?

Reply:
Sorry, was meant to be placed on a comment about the maximum number of redirects allowed being 30 instead of 20.

Reply:
Not sure. 20 seems reasonable given this: https://stackoverflow.com/questions/9384474/in-chrome-how-many-redirects-are-too-many